text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,236 @@
+statically to some sane value. Parachains should prefer to have a static value that makes sense for their usecase which can be changed by governance at some future point. Changing the value dynamically can be a friction point. It will work out fine to decrease the value to build more into the present. But if the value is increased to build more into the future, a relay chain block will be skipped.
+
+Considering `para_assigned_cores` is a sorted vec of core indices assigned to a parachain at the
+specified claim queue offset, validators will determine the committed core index like this:
+
+```
+let assigned_core_index = core_selector % para_assigned_cores.len();
+let committed_core_index = para_assigned_cores[assigned_core_index];
+```
+
+
+### Polkadot Primitive changes
+
+#### New [CandidateDescriptor](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/primitives/src/v7/mod.rs#L482)
+
+- reclaim 32 bytes from `collator: CollatorId` and 64 bytes from `signature: CollatorSignature` and rename to `reserved1`  and `reserved2` fields.
+- take 2 bytes from `reserved1` for a new `core_index: u16` field.  
+- take 4 bytes from `reserved2` for a new `session_index: u32` field.
+- the `reserved1` and `reserved2` fields are zeroed
+
+Thew new primitive will look like this:
+```
+pub struct CandidateDescriptorV2<H = Hash> {
+	/// The ID of the para this is a candidate for.
+	para_id: Id,
+	/// The hash of the relay-chain block this is executed in the context of.
+	relay_parent: H,
+	/// The core index where the candidate is backed.
+	core_index: CoreIndex,
+	/// The session index in which the candidate is backed.
+	session_index: SessionIndex,
+	/// Reserved bytes.
+	reserved1: [u8; 26],
+	/// The blake2-256 hash of the persisted validation data. This is extra data derived from
+	/// relay-chain state which may vary based on bitfields included before the candidate.
+	/// Thus it cannot be derived entirely from the relay-parent.
+	persisted_validation_data_hash: Hash,
+	/// The blake2-256 hash of the PoV.
+	pov_hash: Hash,
+	/// The root of a block's erasure encoding Merkle tree.
+	erasure_root: Hash,
+	/// Reserved bytes.
+	reserved2: [u8; 64],
+	/// Hash of the para header that is being generated by this candidate.
+	para_head: Hash,
+	/// The blake2-256 hash of the validation code bytes.
+	validation_code_hash: ValidationCodeHash,
+}
+```
+
+In future format versions, parts of the `reserved1` and `reserved2` bytes can be used to include additional information in the descriptor.
+
+#### Versioned `CandidateReceipt` and `CommittedCandidateReceipt` primitives:
+
+We want to decouple the actual representation of the `CandidateReceipt` from the higher level code. This should make it easier to implement future format versions of this primitive. To hide the logic of versioning the descriptor fields will be private and accessor methods are provided.
+
+``` 
+pub enum VersionedCandidateReceipt {
```

Comment:
An alternative proposal is to only make above data structures with all fields private. Then you implement the accessor functions like this (pseudo code):

```rust
impl CandidateDescriptor {
  pub fn core_index(&self) -> Option<CoreIndex> {
    if self.reserved2 == 0 {
      Some(self.core_index)
    } else {
      None
    }
 }
 // ... same for SessionIndex
 
  pub fn signature(&self) -> Option<&CollatorSignature> {
    if self.reserved2 == 0 {
      None
    } else {
      Some(&self.reserved2)
    }
  } 
  // similar for CollatorId, but need to be reconstructed from SessionIndex, CoreIndex and reserved1 fields.
}
```

This way even internal representations stay binary compatible and we also don't need a manual `Decode` instance. Not clear what is better: manual decode or leaving the "decode" to accessor functions ... so just food for thought really.

Reply:
This is pretty much similar to what I am currently implementing, except we still wrap the receipt versions in this enum which exposes accessors and handles the versioning logic in the `Decode` implementation.

Reply:
We could use the derived `Decode` and always decode it as latest version. At least for the runtime, the collator singature and id are not even used, so decoding in accessor functions is only needed on the node. It's not really decoding, more like concatenating the bytes from `core_index`, `session_index`, `reserved1` and `reserved2` can be easily used to construct the signature.

This might cause way less changes than using the `enum`.

Reply:
I like the idea of not having an Enum and having accessor functions

Reply:
Seems the only downsides are that decoding these receipts will show junk in v2 fields if the descriptor is v1.